### PR TITLE
Demzne 1213 endpoint dao user status   for badge details

### PR DESCRIPTION
**Pull request by sgurmani**
It is on halt because we don't know yet how much time it will take in real-time to come from the history tools-> database.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -24,5 +24,33 @@
...
+      ],
+      "import/prefer-default-export": "off",
+      "no-nested-ternary": "off",
+      "@typescript-eslint/no-unused-vars": [
+        "error",
+        {
+          "argsIgnorePattern": "^_"
+        }
+      ]
+
```
Comment (sgurmani) [inline .eslintrc.json:55]: Added these config because it would work without additional extension installed on the vscode or whatever.

File: src/endpoints/user-status/routes/user-status.route.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ *
+ */
+export class GetUserStatusRoute extends GetRoute {
+	public static create(handler: RouteHandler) {
+		return new GetUserStatusRoute(handler);
+	}
+
+	private constructor(handler: RouteHandler) {
+		super('/v1/eosdac/userstatus', handler, {
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/routes/user-status.route.ts:24]: I suggest defining the route as `/v1/eosdac/user_status`

File: src/config/config.types.ts [hunk 3]
```diff
@@ -43,3 +43,6 @@ export type DocsConfig = {
 export type DACConfig = {
 	nameCache: Map<string, DacDirectory>;
 };
+export type DACConfigArray = {
```
Comment (ahmadtirmazi) [inline src/config/config.types.ts:46]: I believe, we do not need to introduce a separate data type `DACConfigArray`. 
We can populate all the DACs configs in the already available map here at [Line 43](https://github.com/Alien-Worlds/eosdac-api/pull/50/files?diff=unified&w=0#diff-e9ca1bd3c1029dcafec9c13165847fca0815228497b5bf62cc9b076aa8d0a101R43).

The `dacId` is already being used as a key in the existing map.

File: src/config/app-config.ts [hunk 4]
```diff
@@ -43,6 +44,11 @@ export default class AppConfig implements Config {
 
 		public readonly dac: DACConfig = {
 			nameCache: new Map(),
+		},
+
+		public readonly setOfDacs: DACConfigArray = {
```
Comment (ahmadtirmazi) [inline src/config/app-config.ts:49]: Same goes here, we won't require this.

File: src/endpoints/user-status/domain/userstatus.controller.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import {
+	DacDirectory,
+	IndexWorldsContract,
+} from '@alien-worlds/eosdac-api-common';
+import { Failure, injectable, Result } from '@alien-worlds/api-core';
+import { config } from '@config';
+import { GetUserStatusInput } from './models/get-user-status.input';
+import { GetUserStatusOutput } from './models/get-user-status.output';
+import { inject } from 'inversify';
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/domain/userstatus.controller.ts:9]: We are trying to depend more on the api-core and less on external packages.

TLDR; please import this from `@alien-worlds/api-core` instead of `inversify` everywhere.

File: src/endpoints/user-status/data/dtos/userstatus.dto.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+export type GetUserStatustDto = {
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/data/dtos/userstatus.dto.ts:1]: Conventionally, we use the suffix `RequestDto`.

Can we rename this to `GetUserStatustRequestDto`?

File: src/endpoints/user-status/data/dtos/userstatus.dto.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	walletId: string;
+};
+
+export enum DAOUserStatusType {
+	EXPLORER = 'Explorer',
+	MEMBER = 'Member',
+	CANDIDATE = 'Candidate',
+	CUSTODIAN = 'Custodian',
+}
+export type DaoUserStatuses = {
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/data/dtos/userstatus.dto.ts:11]: Probably singular name should be more meaningful? i.e. `DAOUserStatus`

File: src/endpoints/user-status/data/dtos/userstatus.dto.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	name: string;
+	status: DAOUserStatusType;
+};
+export enum ERROR_MESSAGE_TYPE {
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/data/dtos/userstatus.dto.ts:15]: I understand the need here to have a custom error type. 
Probably, we should define it as a separate error type similar to what we have here

https://github.com/Alien-Worlds/eosdac-api/blob/master/src/common/api/domain/errors/limit-exceeded.error.ts

File: src/endpoints/user-status/domain/models/get-user-status.input.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import { GetUserStatustDto } from '../../data/dtos/userstatus.dto';
+import { Request } from '@alien-worlds/api-core';
+/**
+ * @class
+ */
+export class GetUserStatusInput {
+	/**
+	 *
+	 * @param {CandidateRequestDto} dto
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/domain/models/get-user-status.input.ts:9]: Just a minor thing, probably need to update the param type here to `GetUserStatustDto`.

File: src/endpoints/user-status/domain/use-cases/get-user-status.use-case.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+   * @returns {Promise<Result<DaoUserStatus[]>>}
+   */
+  public async execute(
+    walletId: string,
+    dacs: DacDirectory[]
+  ): Promise<Result<DaoUserStatus[]>> {
+    const result: DaoUserStatus[] = [];
+    let status: DAOUserStatusType = DAOUserStatusType.EXPLORER;
+    for (const dac of dacs) {
+      status = DAOUserStatusType.EXPLORER;
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/domain/use-cases/get-user-status.use-case.ts:45]: May be we define this variable here, instead of Line 43?
Since, it is being assigned with every iteration so shouldn't affect.

File: src/endpoints/user-status/domain/use-cases/get-user-status.use-case.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+      const { content: custodian, failure: getCustodianFailure } =
+        await this.getCustodianUseCase.execute(dac.dacId, walletId);
+
+      if (
+        getCustodianFailure &&
+        getCustodianFailure.error.message !== ERROR_MESSAGE_TYPE.NOTFOUND
```
Comment (ahmadtirmazi) [inline src/endpoints/user-status/domain/use-cases/get-user-status.use-case.ts:52]: Here we may check the type of error instead of relying on the error message.

**Review by ahmadtirmazi — CHANGES_REQUESTED** [review]

**Comment by ahmadtirmazi** [comment]
@sgurmani, probably also consider writing an integration test for this new endpoint.
Though, I am working on developing a new approach about it but the use cases that you will identify will be valuable.
